BTrDB.cs

```diff
@@ -1,131 +1,732 @@
+        /// </summary>
+        internal Stream(BTrDB db, V5Api.StreamDescriptor desc)
+        {
+            this.db = db;
+            id = new Guid(desc.Uuid.ToByteArray());
+            collection = desc.Collection;
+            tags = desc.Tags.ToDictionary(e => e.Key, e => e.Val?.Value);
+            annotations = desc.Annotations.ToDictionary(e => e.Key, e => e.Val?.Value);
+            propertyVersion = desc.PropertyVersion;
         }
 
+        /// <summary>
+        /// Check if the stream exists
+        /// </summary>
+        /// <returns>true if the stream exists, false otherwise</returns>
         public bool Exists()
         {
-            throw new NotImplementedException("not implemented");
+            var p = new V5Api.StreamInfoParams
+            {
+                Uuid = ByteString.CopyFrom(id.ToByteArray()),
+            };
+            var results = db.client.StreamInfo(p);
+            if (results.Stat != null)
+            {
+                if (results.Stat.Code == 404)
+                {
+                    return false;
+                }
+                throw new BTrDBException(results.Stat);
+            }
+            return true;
         }
 
-        public IEnumerable<RawPoint> RawValues(long start, long end, ulong version = BTrDB.LatestVersion)
+        /// <summary>
+        /// Refresh the metadata (tags and annotations)
+        /// </summary>
+        private void RefreshMeta()
         {
-            throw new NotImplementedException("not implemented");
+            V5Api.StreamInfoParams p = new V5Api.StreamInfoParams
+            {
+                Uuid = ByteString.CopyFrom(id.ToByteArray()),
+            };
+            V5Api.StreamInfoResponse res = db.client.StreamInfo(p);
+            if (res.Stat != null) throw new BTrDBException(res.Stat);
+            collection = res.Descriptor_.Collection;
+            tags = res.Descriptor_.Tags.ToDictionary(e => e.Key, e => e.Val?.Value);
+            annotations = res.Descriptor_.Annotations.ToDictionary(e => e.Key, e => e.Val?.Value);
+            propertyVersion = res.Descriptor_.PropertyVersion;
         }
 
-        public IEnumerable<StatPoint> AlignedWindows(long start, long end, ushort pointWidth, ulong version = BTrDB.LatestVersion)
+        /// <summary>
+        /// Query the raw values in the stream
+        /// </summary>
+        /// <param name="start">The (inclusive) starting timestamp in nanoseconds since the Unix Epoch (Jan 1st 1970)</param>
+        /// <param name="end">The (exclusive) ending timestamp in nanoseconds since the Unix Epoch (Jan 1st 1970)</param>
+        /// <param name="version">The version of the stream to query. If unspecified, the latest version will be used</param>
+        /// <returns>A Cursor of RawValues that can be enumerated. <see cref="ICursor"/></returns>
+        public ICursor<RawPoint> RawValues(long start, long end, ulong version = BTrDB.LatestVersion)
```

Comment:
I very much like this ICursor interface … I know you did extra work to get the Wait() in there and I think it's going to pay off.

Reply:
That ICursor was the result of many iterations...